cluster_experiments/experiment_analysis.py

```diff
@@ -301,6 +301,92 @@ def from_config(cls, config):
         )
 
 
+class PairedTTestClusteredAnalysis(ExperimentAnalysis):
+    """
+    Class to run paired T-test analysis on aggregated data
+
+    Arguments:
+        cluster_cols: list of columns to use as clusters
+        target_col: name of the column containing the variable to measure
+        treatment_col: name of the column containing the treatment variable
+        treatment: name of the treatment to use as the treated group
+        strata: column to generate 2 groups (a and b)
+
+    Usage:
+
+    ```python
+    from cluster_experiments.experiment_analysis import TTestClusteredAnalysis
+    import pandas as pd
+
+    df = pd.DataFrame({
+        'x': [1, 2, 3, 4, 0, 0, 1, 1],
+        'treatment': ["A", "B", "A", "B"] * 2,
+        'cluster': [1, 2, 3, 4, 1, 2, 3, 4],
+    })
+
+    PairedTTestClusteredAnalysis(
+        strata=df['cluster'],
+        target_col=df['x'],
+    ).get_pvalue(df)
+    ```
+    """
+
+    def __init__(
+        self,
+        cluster_cols: List[str],
+        strata_cols: List[str] = None,
+        target_col: str = "target",
+        treatment_col: str = "treatment",
+        treatment: str = "B",
+    ):
+        self.strata_cols = cluster_cols if strata_cols is None else strata_cols
+        self.target_col = target_col
+        self.treatment = treatment
+        self.treatment_col = treatment_col
+        self.cluster_cols = cluster_cols
+
+    def analysis_pvalue(self, df: pd.DataFrame, verbose: bool = False) -> float:
+        """Returns the p-value of the analysis
+        Arguments:
+            df: dataframe containing the data to analyze
+            verbose (Optional): bool, prints the regression summary if True
+        """
+
+        df_grouped = df.groupby(
+            self.cluster_cols + [self.treatment_col], as_index=False
+        )[self.target_col].mean()
+
+        if (
+            df_grouped[self.treatment_col].value_counts()[0]
+            != df_grouped[self.treatment_col].value_counts()[1]
+        ):
+            warnings.warn(
+                "groups don't have same number of observations"
+            )  # todo raise count as well?
+
+        df_pivot = df_grouped.pivot_table(
+            columns=self.treatment_col,
+            index=self.strata_cols,
```

Comment:
in this case, you are using strata_cols (or cluster_cols) as the identifier of the pair, right? What if I have cities as strata_cols and couriers as cluster_cols ? In that case, should the identifier of the pair be cluster_cols + strata_cols in the most generic scenario?

Reply:
hm not sure. I think we should concatenate them because the strata can be different than the cluster_cols in case you  want to compare different things. In your example, if I leave only cities in strata, every row will be a city and it will average all couriers. If you have cluster_cols + strata_cols in strata every row will be compararing the courier level, and  each courier will be either in A or B, right?  For example :

`
def test_paired_ttest():
    "This test make sure that pivot table works as expected (inside get_pvalue) and that paired t test returns a possible value"

    analysis_df = pd.DataFrame(
        {
            "cluster": ["ES"] * 4 + ["IT"] * 4 + ["PL"] * 4 + ["RO"] * 4,
            "date": ["2022-01-01", "2022-01-02", "2022-01-03", "2022-01-04"] * 4,
            "treatment": ["A", "B"] * 8,
            "target": [0.01] * 16,
        }
    )
    # Changing just one observation so we have a p value
    analysis_df.loc[1, "target"] = 0.001

    analyser = PairedTTestClusteredAnalysis(
        cluster_cols=["cluster", "date"], strata_cols=["cluster", "date"]
    )  # todo test passing no strata, test passing many strata

    assert 1 >= analyser.get_pvalue(analysis_df) >= 0, "p value is null or inf"`

Will return:
![image](https://user-images.githubusercontent.com/33734080/216026363-6f4d5a3f-b28d-477d-926b-b09b2dca0d39.png)

Make sense? Maybe it doesn't 😆

Reply:
hmm, I think we should clarify the meaning of strata_cols then. It's not fully clear to me what role does this column has.. My assumption is that cluster_cols and strata_cols are independent one from each other. What I see in the code is that:
- cluster_cols manages the level of pre-aggregation of the metric -> OK
- strata_cols manages a further level of aggregation of the metric, by deciding the level of granularity of the pivot table that puts paired items one next to each other, after having already grouped the rows at the cluster-treatment level. -> Here I'm a bit confused.

Where would I use strata_cols that is different from cluster_cols, and how, in a practical example?

Reply:
Thanks Ludo for comments and David for summarising it. I think the final conclusion is that strata_cols should always be either a subset or equals to cluster_cols

Reply:
I disagree @ludovico-lanni. In switchbacks for example you always want to cluster by date, but you will add to the strata cols

Reply:
ahhh, ok now I get the subset part, I misunderstood it! Ok I see what you mean, I think this serves a specific use case like the one David described.. makes sense to me! Sorry for the late catch 🆙